pkg/auth/dependency/webapp/state_provider.go

```diff
@@ -0,0 +1,103 @@
+package webapp
+
+import (
+	"net/http"
+)
+
+//go:generate mockgen -source=state_provider.go -destination=state_provider_mock_test.go -package webapp
+
+type StateStore interface {
+	Get(id string) (*State, error)
+	Set(state *State) error
+}
+
+// StateProvider does not implicitly generate new state for invalid or expired sid.
+// If sid is present, then it must be valid.
+type StateProvider interface {
+	// PersistError assumes id exists.
+	PersistError(id string, inputError error)
+	PersistState(r *http.Request, inputError error)
+	// MayRestoreState may return nil, nil.
+	MayRestoreState(r *http.Request) (state *State, err error)
```

Comment:
Actually, I wonder whether it's conventional to return `nil, nil` in Go? I seldom seen this behavior from my limitation experience with Go. I've written some code with this behavior before, and it seems getting confusing since we need to remember whether `nil` is possible for some function.

What's your opinion on this?